stacker/actions/diff.py

```diff
@@ -0,0 +1,146 @@
+import logging
+
+from .. import exceptions
+from ..plan import COMPLETE, SKIPPED, Plan
+from . import build
+import difflib
+import json
+import sys
+
+logger = logging.getLogger(__name__)
+
+
+class Action(build.Action):
+    """ Responsible for diff'ing CF stacks in AWS and on disk
+
+    Generates the build plan based on stack dependencies (these dependencies
+    are determined automatically based on references to output values from
+    other stacks).
+
+    The plan is then used to pull the current CloudFormation template from
+    AWS and compare it to the generated templated based on the current
+    config.
+    """
+
+    def _diff_parameters(self, old_params, new_params):
+        """Compares the old vs. new parameters and returns a 'diff'
+
+        Args:
+            old_params(dict): old paramters
+            new_params(dict): new parameters
+
+        Returns:
+            int: number of changed entries
+            string: "diff" of all parameters and how they changed
+        """
+        output = []
+        count = 0
+        for k, v in old_params.iteritems():
+            new_val = new_params.get(k, "_Undef_")
+            if str(v) != str(new_val):
+                output.append("CHANGED   %s: %s => %s\n" % (k, v, new_val))
+                count += 1
+            else:
+                output.append("NO CHANGE %s: %s\n" % (k, v))
+
+        for k, v in new_params.iteritems():
+            if not k in old_params:
+                output.append("NEW       %s: %s\n" % (k, v))
+                count += 1
+
+        return [count, "".join(output)]
+
+    def _normalize_json(self, template):
+        """Normalizes our template for diffing
+
+        Args:
+            template(str): json string representing the template
+
+        Returns:
+            list: json representation of the parameters
+        """
+        obj = json.loads(template)
+        json_str = json.dumps(obj, sort_keys=True, indent=4)
+        result = []
+        lines = json_str.split("\n")
+        for line in lines:
+            result.append(line + "\n")
+        return result
+
+    def _diff_stack(self, stack, **kwargs):
+        """Handles the diffing a stack in CloudFormation vs our config"""
+        if not build.should_submit(stack) or not build.should_update(stack):
+            return SKIPPED
+
+        # get the current stack template & params from AWS
+        try:
+            [old_template, old_params] = self.provider.get_stack_info(
+                stack.fqn)
+        except exceptions.StackDoesNotExist:
+            old_template = None
+            old_params = {}
+
+        # generate our own template & params
+        new_template = stack.blueprint.rendered
+        resolved_parameters = self._resolve_parameters(stack.parameters,
```

Comment:
i would also roll these into one method: https://github.com/remind101/stacker/blob/master/stacker/actions/build.py#L173-L178 which you could call here instead of copying over the same workflow which could change in the future